overlord/snapshotstate/backend/backend.go

```diff
@@ -378,7 +387,54 @@ func Save(ctx context.Context, id uint64, si *snap.Info, cfg map[string]interfac
 
 var isTesting = snapdenv.Testing()
 
-func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer, username string, entry, dir string) error {
+type addDirToZipOptions struct {
+	ExcludePaths []string `yaml:"exclude,omitempty"`
+}
+
+func readSnapshotYaml(si *snap.Info, opts *addDirToZipOptions) error {
+	file, err := osOpen(filepath.Join(si.MountDir(), "meta", "snapshots.yaml"))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		} else {
+			return nil
+		}
+	}
+	defer file.Close()
+
+	if err := yaml.NewDecoder(file).Decode(opts); err != nil {
+		return fmt.Errorf("cannot read snapshot manifest: %v", err)
+	}
+
+	// Validate the exclude list; note that this is an *exclusion* list, so
+	// even if the manifest specified paths starting with ../ this would not
+	// cause tar to navigate into those directories and pose a security risk.
+	// Still, let's have a minimal validation on them being sensible.
+	validFirstComponents := []string{
+		"$SNAP_DATA", "$SNAP_COMMON", "$SNAP_USER_DATA", "$SNAP_USER_COMMON",
+	}
+	for _, excludePath := range opts.ExcludePaths {
+		firstSlash := strings.IndexRune(excludePath, '/')
+		var firstComponent string
+		if firstSlash < 0 {
+			firstComponent = excludePath
+		} else {
+			firstComponent = excludePath[:firstSlash]
+		}
+		if !strutil.ListContains(validFirstComponents, firstComponent) {
+			return fmt.Errorf("snapshot exclude path must start with one of %q (got: %q)", validFirstComponents, excludePath)
+		}
+
+		cleanPath := filepath.Clean(excludePath)
+		if cleanPath != excludePath {
+			return fmt.Errorf("snapshot exclude path not clean: %q", excludePath)
```

Comment:
This is the validation :-)

Yes, you are right, that if the snapshots.yaml file is invalid, then the snapshot operation will fail. We might want to not make it fatal and act as if it was not there, maybe. @pedronis, do you have an opinion?

Reply:
Fwiw, I think we should still do a warning or something here instead of a hard error. It might be that we make the validation more strict over time and people might have snapshot.yaml installed that passed the validation back then but not anymore (hope this makes sense?)